### DISCUSSION TITLE
ESP32-PICO-V3-02 + GC9A01 (ESP32_GENERIC,SPIRAM)

### DISCUSSION THREAD

**Discussion by sam0910** [lvgl-micropython/lvgl_micropython#102] [Board Specific]
> [!NOTE]
> ESP32 with SPIRAM throws IRAM overflow error when compile.
> Solutions are, 
> 1. if you are using ESP32 revision ECO3, 
> add `CONFIG_ESP32_REV_MIN_3=y` into `ports/esp32/boards/sdkconfig.spiram`
> This will save 10Kb of IRAM and it is enough for compiling with `lcd_bus`
>
> 2. Disable some IRAM functions (insert to 'sdkconfig.spiram') , 
> `CONFIG_ESP_PANIC_HANDLER_IRAM=n `
> `CONFIG_LWIP_IRAM_OPTIMIZATION=n `
> `CONFIG_SPI_MASTER_ISR_IN_IRAM=n .. etc`
> There will be trade off for this, please see the [Optimizing IRAM Usage](https://docs.espressif.com/projects/esp-idf/en/stable/esp32/api-guides/performance/ram-usage.html#optimizing-iram-usage).
>  

Build script
`python3 make.py esp32 submodules clean mpy_cross BOARD=ESP32_GENERIC BOARD_VARIANT=SPIRAM DISPLAY=gc9a01`

```
import lcd_bus
import task_handler  # NOQA
import time  # NOQA
from micropython import const

# display settings
WIDTH = const(240)
HEIGHT = const(240)
BUFFER_SIZE = WIDTH * HEIGHT // 4

MOSI = const(13)  
CLK = const(14) 
CS = const(15) 
DC = const(25)  

FREQ = const(60000000)

from machine import SPI  
lcd_spi_bus = SPI.Bus(host=1, sck=CLK, mosi=MOSI, miso=-1)

display_bus = lcd_bus.SPIBus(
    spi_bus=lcd_spi_bus,
    dc=DC,
    freq=FREQ,
    cs=CS,
    spi_mode=0,
)

fb1 = display_bus.allocate_framebuffer(BUFFER_SIZE, lcd_bus.MEMORY_INTERNAL | lcd_bus.MEMORY_DMA)
fb2 = display_bus.allocate_framebuffer(BUFFER_SIZE, lcd_bus.MEMORY_INTERNAL | lcd_bus.MEMORY_DMA)

import gc9a01  # NOQA
import lvgl as lv  # NOQA

lv.init()

display = gc9a01.GC9A01(
    data_bus=display_bus,
    frame_buffer1=fb1,
    frame_buffer2=fb2,
    display_width=WIDTH,
    display_height=HEIGHT,
    color_space=lv.COLOR_FORMAT.RGB565,
    color_byte_order=gc9a01.BYTE_ORDER_BGR,
    rgb565_byte_swap=True,
)


display.init()
display.invert_colors()
display.set_rotation(lv.DISPLAY_ROTATION._90)

th = task_handler.TaskHandler(5)

scrn = lv.screen_active()
scrn.set_style_bg_color(lv.color_hex(0x000000), 0)  # 0x000000 , 0x0000FF

label = lv.label(scrn)
label.set_text("HELLO WORLD!")
label.set_style_text_color(lv.color_hex(0x0000FF), 0)
label.align(lv.ALIGN.CENTER, 0, 0)
```

**Comment by kdschlosser** [comment]
All of those config settings are able to be passed from the command line.

```
 python3 make.py esp32 submodules clean mpy_cross BOARD=ESP32_GENERIC BOARD_VARIANT=SPIRAM DISPLAY=gc9a01 CONFIG_ESP32_REV_MIN_3=y CONFIG_ESP_PANIC_HANDLER_IRAM=n CONFIG_LWIP_IRAM_OPTIMIZATION=n CONFIG_SPI_MASTER_ISR_IN_IRAM=n
```

you don't have to edit any files to add them.

**Comment by kdschlosser** [comment]
also when dealing with the SPI Bus I recommend using a buffer size that is 1/10th of the total display size. `width * height * color depth in bytes / 10` we have found this is the optimum setting performance wise.

**Comment by kdschlosser** [comment]
the other thing is that setting the task handler to update every 5 milliseconds is a but much and it will consume a lot of your processor time doing that. I recommend leaving it at it's default of 33 milliseconds.

**Comment by sam0910** [comment]
Thank you so much for the review!, and that is awesome!

**Comment by kdschlosser** [comment]
no worries m8, just thought you should know those couple of things.
